include/openthread/tcp.h

```diff
@@ -0,0 +1,589 @@
+ * @param[in]  aSendDoneCallback  The new callback.
+ *
+ * @retval OT_ERROR_NONE    Successfully set the callback.
+ * @retval OT_ERROR_FAILED  Failed to set the callback.
+ */
+otError otTcpActiveSetSendDoneCallback(otInstance *aInstance, otTcpActiveSocket *aSocket, otTcpSendDone aSendDoneCallback);
+
+/**
+ * Sets the "send ready" callback on an active TCP socket.
+ *
+ * @param[in]  aInstance           A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket             A pointer to the active TCP socket structure whose callback to set.
+ * @param[in]  aSendReadyCallback  The new callback.
+ *
+ * @retval OT_ERROR_NONE    Successfully set the callback.
+ * @retval OT_ERROR_FAILED  Failed to set the callback.
+ */
+otError otTcpActiveSetSendReadyCallback(otInstance *aInstance, otTcpActiveSocket *aSocket, otTcpSendReady aSendReadyCallback);
+
+/**
+ * Sets the "receive available" callback on an active TCP socket.
+ *
+ * @param[in]  aInstance                  A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket                    A pointer to the active TCP socket structure whose callback to set.
+ * @param[in]  aReceiveAvailableCallback  The new callback.
+ *
+ * @retval OT_ERROR_NONE    Successfully set the callback.
+ * @retval OT_ERROR_FAILED  Failed to set the callback.
+ */
+otError otTcpActiveSetReceiveAvailableCallback(otInstance *aInstance, otTcpActiveSocket *aSocket, otTcpReceiveAvailable aReceiveAvailableCallback);
+
+/**
+ * Sets the "connection lost" callback on an active TCP socket.
+ *
+ * @param[in]  aInstance                A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket                  A pointer to the active TCP socket structure whose callback to set.
+ * @param[in]  aConnectionLostCallback  The new callback.
+ *
+ * @retval OT_ERROR_NONE    Successfully set the callback.
+ * @retval OT_ERROR_FAILED  Failed to set the callback.
+ */
+otError otTcpActiveSetConnectionLostCallback(otInstance *aInstance, otTcpActiveSocket *aSocket, otTcpConnectionLost aConnectionLostCallback);
+
+/**
+ * Binds the active socket to an IP address and port.
+ *
+ * @param[in]  aInstance   A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket     A pointer to the active TCP socket structure to bind.
+ * @param[in]  aSockName   The address and port to which to bind this socket.
+ *
+ * @retval OT_ERROR_NONE    Successfully bound the active TCP socket.
+ * @retval OT_ERROR_FAILED  Failed to bind the active TCP socket.
+ */
+otError otTcpActiveBind(otInstance *aInstance, otTcpActiveSocket *aSocket, const otSockAddr *aSockName);
+
+/**
+ * This enumeration defines flags passed to @p otTcpActiveConnect.
+ */
+enum {
+    OT_TCP_ACTIVE_CONNECT_FAST_OPEN =    1 << 0,
+    OT_TCP_ACTIVE_CONNECT_NO_FAST_OPEN = 1 << 1,
```

Comment:
I'm in favor of making this change, but it requires broad consensus that we want fast open by default. I know @StuartCheshire and I would both like to do this (and it seems you would like to as well!) but I wasn't sure if others were already on board with making fast open the default. If others can share their thoughts, that would be helpful.

How about this: I'll leave this comment open for a day or so, and if nobody objects, I'll merge in your suggested change.

Reply:
Actually, I went ahead and pushed an update to include this fix and update the comment description for `otTcpActiveConnect` to clarify that TCP Fast Open is the default behavior. I'm leaving this conversation open to give others a chance to object to this, but I'll close the conversation in a couple of days if nobody objects.

Reply:
I think it's better to use TFO by default since responsiveness matters a lot in home nets.
As long as TFO is widely supported by Host TCP implementations, we should use TFO by default.

Reply:
I agree with making TFO default. As @StuartCheshire mentioned, we want to encourage good behavior.

Reply:
OK! It sounds like we have broad agreement on this. I've already pushed the update, so I'll go ahead and resolve the conversation now.